src/snowflake/snowpark/session.py

```diff
@@ -1149,27 +1195,162 @@ def _upload_unsupported_packages(
             )
 
             # Zip and add to stage
-            zip_file = f"{IMPLICIT_ZIP_FILE_NAME}_{generate_random_alphanumeric(5)}.zip"
+            environment_signature: str = get_signature(packages)
+            zip_file = f"{IMPLICIT_ZIP_FILE_NAME}_{environment_signature}.zip"
             zip_path = os.path.join(tmpdir, zip_file)
             zip_directory_contents(target, zip_path)
 
-            # TODO: Make this upload non-lazy and allow custom stages / zip path names
-            self.add_import(zip_path)
+            # Add packages to stage
+            stage_name = self.get_session_stage()
+
+            if persist_path:
+                # Switch the stage used for storing zip file.
+                stage_name = persist_path
+
+                # Download metadata dictionary using the technique mentioned here: https://docs.snowflake.com/en/user-guide/querying-stage
+                metadata_file = f"{ENVIRONMENT_METADATA_FILE_NAME}.txt"
+                normalized_metadata_path = normalize_remote_file_or_dir(
+                    f"{persist_path}/{metadata_file}"
+                )
+                metadata = {
+                    row[0]: row[1] if row[1] else []
+                    for row in (
+                        self.sql(
+                            f"SELECT t.$1 as signature, t.$2 as packages from {normalized_metadata_path} t"
+                        )._internal_collect_with_tag()
+                    )
+                }
+                _logger.info(f"METADATA: {metadata}")
+
+                # Add a new enviroment to the metadata, avoid commas while storing list of dependencies because commas are treated as default delimiters.
+                metadata[environment_signature] = "|".join(
+                    [
+                        str(requirement)
+                        for requirement in supported_dependencies + new_dependencies
+                    ]
+                )
+                metadata_local_path = os.path.join(
+                    self._tmpdir_handler.name, metadata_file
+                )
+                with open(metadata_local_path, "w") as file:
+                    for key, value in metadata.items():
+                        file.write(f"{key},{value}\n")
+
+                # Upload metadata file to stage
+                # Note that the metadata file is not compressed, only the zip files are.
+                self._conn.upload_file(
+                    path=normalize_local_file(metadata_local_path),
+                    stage_location=normalize_remote_file_or_dir(stage_name),
+                    compress_data=False,
+                    overwrite=True,
+                )
+
+            self._conn.upload_file(
+                path=normalize_local_file(zip_path),
+                stage_location=normalize_remote_file_or_dir(stage_name),
+                compress_data=True,
+                overwrite=True,
```

Comment:
Don't you think you should use `overwrite=reinstall_packages`. This is allow users to populate zip file if is it not present. In case file is present, it will skip the overwrite.